tests/jlm/util/TestTarjanScc.cpp

```diff
@@ -300,3 +300,50 @@ TestDiamondChainWithBackEdge()
 JLM_UNIT_TEST_REGISTER(
     "jlm/util/TestTarjanScc-TestDiamondChainWithBackEdge",
     TestDiamondChainWithBackEdge);
+
+// During SCC creation, the function should query a node for its successors at most twice
+static int
+TestVisitEachNodeTwice()
+{
+  const size_t numNodes = 5;
+  std::vector<std::vector<size_t>> successors{
+    { 1, 2 }, // 0's successors
+    { 2, 3 }, // 1's successors
+    { 1, 3 }, // 2's successors
+    {},       // 3's successors
+    { 4 }     // 4's successors
+  };
+  // The graph looks like a 0->(12)->3 diamond with edges between 1 and 2, and a lone node 4
+
+  std::vector<size_t> successorsQueried(numNodes, 0);
+  auto GetSuccessors = [&](size_t node)
+  {
+    JLM_ASSERT(node < numNodes);
+    successorsQueried[node]++;
+    return successors[node];
+  };
+
+  std::vector<size_t> sccIndex;
+  std::vector<size_t> topologicalOrder;
+  auto numSccs = jlm::util::FindStronglyConnectedComponents(
+      numNodes,
+      GetSuccessors,
+      sccIndex,
+      topologicalOrder);
+
+  JLM_ASSERT(numSccs == 4);
+  for (size_t timesQueried : successorsQueried)
+    JLM_ASSERT(timesQueried <= 2);
```

Comment:
So I might be wrong here, as I do not fully know the details by heart, but should this not be `== 1` for each of the nodes, i.e., the successor for a node was only asked for once (and therefore the node was only visited once?

Reply:
If the different nodes have a different number in successorQueried, i.e., 0 or 1, or 2, maybe we can be more precise here and assert for all the nodes, e.g.,  `assert(successorsQueried[0] == 0)`

Reply:
Currently every node is queried for its children exactly twice. Once when it is visited for the first name (and needs to ensure all successors are visited), and once again after all successors have been visited. These two visits can be separated by visits of an arbitrary amount of other nodes, so it has to query at least some nodes twice. Anything more than that is unnecessary.

The usual call stack based dfs places the iterators on the call stack, but I assume that getting the list of successors again is cheaper than adding them to the explicit stack.